### ISSUE TITLE
improper DCE after seemingly-unrelated change

### ISSUE THREAD

**Issue by leeoniya** [rollup/rollup#3555]
@lukastaegert 

- Rollup Version: 2.9.0
- Operating System (or Browser): Windows
- Node Version (if applicable):
- Link to reproduction (IMPORTANT, read below): https://github.com/leeoniya/uPlot/commit/625f6003a50c3f0c5caf2725061661aa0a3ef35f

the relevant changes are minimal (just the uPlot.js file): https://github.com/leeoniya/uPlot/commit/625f6003a50c3f0c5caf2725061661aa0a3ef35f#diff-0e94ed30087b62b2ab95466722c48f84

I have not been able to get a minimal repro in the REPL :(, so hopefully the provided commit is sufficient.

### Expected Behavior

instances of `series.splice()` are retained in the bundle, since they continue to operate on reachable variables both before and after the linked commit.

### Actual Behavior

instances of `series.splice()` are now purged from the bundle:

https://github.com/leeoniya/uPlot/commit/625f6003a50c3f0c5caf2725061661aa0a3ef35f#diff-deb943c7f48de2a8aecf97812918454aL1114-L1127

hopefully something simple (or my mistake).

thanks!

**Comment by leeoniya** [comment]
some minor movement of code assignments works around this issue:

https://github.com/leeoniya/uPlot/commit/abacd02d19403048dbfd30b2004238e203a25598#diff-0e94ed30087b62b2ab95466722c48f84

**Comment by kzc** [comment]
To repro most projects just diff the bundle generated by:
```
    rollup -c rollup.config.js --no-treeshake
```
vs
```
    rollup -c rollup.config.js
```
Then remove code from the first bundle until the test case is reduced...

```
$ rollup -v
rollup v2.10.0
```
```
$ cat bug.js
function f(x) {
    return (x ? [] : ["FAIL"]).map(o => o);
}
var a = f(0);
a.splice(0, 0, "PASS");
console.log(a[0]);
```
Expected:
```
$ cat bug.js | node
PASS
```
```
$ rollup bug.js --silent --no-treeshake | node
PASS
```
Actual:
```
$ rollup bug.js --silent | node
FAIL
```
```
$ rollup bug.js --silent
function f(x) {
    return (x ? [] : ["FAIL"]).map(o => o);
}
var a = f(0);
console.log(a[0]);
```

**Comment by kzc** [comment]
This bug also affects other mutable Array methods.

```
$ cat bug2.js 
function f(x) {
    return (x ? [1, 2] : [3, 4]).map(o => o);
}
var a = f(0);
a.splice(0, 0, 5);
a.reverse();
console.log(a);
```
Expected:
```
$ cat bug2.js | node
[ 4, 3, 5 ]
```
Actual:
```
$ cat bug2.js | dist/bin/rollup --silent | node
[ 3, 4 ]
```
```
$ cat bug2.js | dist/bin/rollup --silent
function f(x) {
    return (x ? [1, 2] : [3, 4]).map(o => o);
}
var a = f(0);
console.log(a);
```
Possible fix:
```
--- a/src/ast/values.ts
+++ b/src/ast/values.ts
@@ -106,7 +106,7 @@ export class UnknownArrayExpression implements ExpressionEntity {
                context: HasEffectsContext
        ) {
                if (path.length === 1) {
-                       return hasMemberEffectWhenCalled(arrayMembers, path[0], this.included, callOptions, context);
+                       return hasMemberEffectWhenCalled(arrayMembers, path[0], true, callOptions, context);
                }
                return true;
        }
@@ -323,7 +323,7 @@ export class UnknownObjectExpression implements ExpressionEntity {
                context: HasEffectsContext
        ) {
                if (path.length === 1) {
-                       return hasMemberEffectWhenCalled(objectMembers, path[0], this.included, callOptions, context);
+                       return hasMemberEffectWhenCalled(objectMembers, path[0], true, callOptions, context);
                }
                return true;
        }
```
With fix:
```
$ cat bug2.js | dist/bin/rollup --silent | node
[ 4, 3, 5 ]
```

**Comment by lukastaegert** [comment]
Thanks for reporting an special thanks to @kzc for that minimal repro, that was really helpful! Fix at #3559
